### ISSUE TITLE
Extremely slow detach/attach via controller-manager (csi_attacher)

### ISSUE THREAD

**Issue by kvaps** [kubernetes/kubernetes#86281]
**What happened**:

Hi, after upgrade to 1.17 we started having the next trouble, the pod migration is taking ages if any csi-volume attached. Example:

**controller-manager v1.16.4** *(17 seconds)*:
```
I1215 10:26:19.538523       1 event.go:274] Event(v1.ObjectReference{Kind:"StatefulSet", Namespace:"hosting", Name:"hc1-md42-s1", UID:"7a67a484-39ac-11e9-bffa-001999d764e2", APIVersion:"apps/v1", ResourceVersion:"2118927898", FieldPath:""}): type: 'Normal' reason: 'SuccessfulCreate' create Pod hc1-md42-s1-0 in StatefulSet hc1-md42-s1 successful
W1215 10:26:19.735797       1 reconciler.go:328] Multi-Attach error for volume "hosting-vol-data-mariadb-hc1-md42-1" (UniqueName: "kubernetes.io/csi/linstor.csi.linbit.com^hosting-vol-data-mariadb-hc1-md42-1") from node "m11c15" Volume is already exclusively attached to node m8c26 and can't be attached to another
I1215 10:26:19.735933       1 event.go:274] Event(v1.ObjectReference{Kind:"Pod", Namespace:"hosting", Name:"hc1-md42-s1-0", UID:"b85722d3-4888-41ff-bb1b-681082a6ba97", APIVersion:"v1", ResourceVersion:"2118928435", FieldPath:""}): type: 'Warning' reason: 'FailedAttachVolume' Multi-Attach error for volume "hosting-vol-data-mariadb-hc1-md42-1" Volume is already exclusively attached to one node and can't be attached to another
I1215 10:26:33.510132       1 operation_generator.go:1474] Verified volume is safe to detach for volume "hosting-vol-data-mariadb-hc1-md42-1" (UniqueName: "kubernetes.io/csi/linstor.csi.linbit.com^hosting-vol-data-mariadb-hc1-md42-1") on node "m8c26" 
I1215 10:26:35.368314       1 operation_generator.go:558] DetachVolume.Detach succeeded for volume "hosting-vol-data-mariadb-hc1-md42-1" (UniqueName: "kubernetes.io/csi/linstor.csi.linbit.com^hosting-vol-data-mariadb-hc1-md42-1") on node "m8c26" 
I1215 10:26:35.388870       1 reconciler.go:288] attacherDetacher.AttachVolume started for volume "hosting-vol-data-mariadb-hc1-md42-1" (UniqueName: "kubernetes.io/csi/linstor.csi.linbit.com^hosting-vol-data-mariadb-hc1-md42-1") from node "m11c15" 
I1215 10:26:36.458511       1 operation_generator.go:390] AttachVolume.Attach succeeded for volume "hosting-vol-data-mariadb-hc1-md42-1" (UniqueName: "kubernetes.io/csi/linstor.csi.linbit.com^hosting-vol-data-mariadb-hc1-md42-1") from node "m11c15" 
I1215 10:26:36.458643       1 event.go:274] Event(v1.ObjectReference{Kind:"Pod", Namespace:"hosting", Name:"hc1-md42-s1-0", UID:"b85722d3-4888-41ff-bb1b-681082a6ba97", APIVersion:"v1", ResourceVersion:"2118928435", FieldPath:""}): type: 'Normal' reason: 'SuccessfulAttachVolume' AttachVolume.Attach succeeded for volume "hosting-vol-data-mariadb-hc1-md42-1" 
```

**controller-manager v1.17.0** *(6 minutes)*:
```
I1214 22:43:30.331528       1 event.go:281] Event(v1.ObjectReference{Kind:"StatefulSet", Namespace:"hosting", Name:"hc1-md42-s1", UID:"7a67a484-39ac-11e9-bffa-001999d764e2", APIVersion:"apps/v1", ResourceVersion:"2114111129", FieldPath:""}): type: 'Normal' reason: 'SuccessfulCreate' create Pod hc1-md42-s1-0 in StatefulSet hc1-md42-s1 successful
W1214 22:43:30.513481       1 reconciler.go:328] Multi-Attach error for volume "hosting-vol-data-mariadb-hc1-md42-1" (UniqueName: "kubernetes.io/csi/linstor.csi.linbit.com^hosting-vol-data-mariadb-hc1-md42-1") from node "m11c15" Volume is already exclusively attached to node m8c29 and can't be attached to another
I1214 22:43:30.513592       1 event.go:281] Event(v1.ObjectReference{Kind:"Pod", Namespace:"hosting", Name:"hc1-md42-s1-0", UID:"4f865b89-49ba-4565-9285-7e6ed1e0ed14", APIVersion:"v1", ResourceVersion:"2114111218", FieldPath:""}): type: 'Warning' reason: 'FailedAttachVolume' Multi-Attach error for volume "hosting-vol-data-mariadb-hc1-md42-1" Volume is already exclusively attached to one node and can't be attached to another
I1214 22:44:35.197812       1 reconciler.go:229] attacherDetacher.DetachVolume started for volume "hosting-vol-data-mariadb-hc1-md42-1" (UniqueName: "kubernetes.io/csi/linstor.csi.linbit.com^hosting-vol-data-mariadb-hc1-md42-1") on node "m8c29" 
I1214 22:45:19.493975       1 operation_generator.go:1297] Verified volume is safe to detach for volume "hosting-vol-data-mariadb-hc1-md42-1" (UniqueName: "kubernetes.io/csi/linstor.csi.linbit.com^hosting-vol-data-mariadb-hc1-md42-1") on node "m8c29" 
E1214 22:46:41.643616       1 csi_attacher.go:202] kubernetes.io/csi: attacher.VolumesAreAttached failed for attach.ID=csi-27bdc8eefc86cffa9c3b14fabe1e89d2e00a78fd37a0e08c48ef91740aec0175: volumeattachments.storage.k8s.io "csi-27bdc8eefc86cffa9c3b14fabe1e89d2e00a78fd37a0e08c48ef91740aec0175" not found
E1214 22:46:41.693657       1 csi_attacher.go:202] kubernetes.io/csi: attacher.VolumesAreAttached failed for attach.ID=csi-27bdc8eefc86cffa9c3b14fabe1e89d2e00a78fd37a0e08c48ef91740aec0175: volumeattachments.storage.k8s.io "csi-27bdc8eefc86cffa9c3b14fabe1e89d2e00a78fd37a0e08c48ef91740aec0175" not found
E1214 22:46:41.743549       1 csi_attacher.go:202] kubernetes.io/csi: attacher.VolumesAreAttached failed for attach.ID=csi-27bdc8eefc86cffa9c3b14fabe1e89d2e00a78fd37a0e08c48ef91740aec0175: volumeattachments.storage.k8s.io "csi-27bdc8eefc86cffa9c3b14fabe1e89d2e00a78fd37a0e08c48ef91740aec0175" not found
E1214 22:46:41.793700       1 csi_attacher.go:202] kubernetes.io/csi: attacher.VolumesAreAttached failed for attach.ID=csi-27bdc8eefc86cffa9c3b14fabe1e89d2e00a78fd37a0e08c48ef91740aec0175: volumeattachments.storage.k8s.io "csi-27bdc8eefc86cffa9c3b14fabe1e89d2e00a78fd37a0e08c48ef91740aec0175" not found
I1214 22:47:09.893665       1 operation_generator.go:470] DetachVolume.Detach succeeded for volume "hosting-vol-data-mariadb-hc1-md42-1" (UniqueName: "kubernetes.io/csi/linstor.csi.linbit.com^hosting-vol-data-mariadb-hc1-md42-1") on node "m8c29" 
I1214 22:47:09.924794       1 reconciler.go:288] attacherDetacher.AttachVolume started for volume "hosting-vol-data-mariadb-hc1-md42-1" (UniqueName: "kubernetes.io/csi/linstor.csi.linbit.com^hosting-vol-data-mariadb-hc1-md42-1") from node "m11c15" 
E1214 22:48:55.443323       1 csi_attacher.go:202] kubernetes.io/csi: attacher.VolumesAreAttached failed for attach.ID=csi-27bdc8eefc86cffa9c3b14fabe1e89d2e00a78fd37a0e08c48ef91740aec0175: volumeattachments.storage.k8s.io "csi-27bdc8eefc86cffa9c3b14fabe1e89d2e00a78fd37a0e08c48ef91740aec0175" not found
E1214 22:48:55.493130       1 csi_attacher.go:202] kubernetes.io/csi: attacher.VolumesAreAttached failed for attach.ID=csi-27bdc8eefc86cffa9c3b14fabe1e89d2e00a78fd37a0e08c48ef91740aec0175: volumeattachments.storage.k8s.io "csi-27bdc8eefc86cffa9c3b14fabe1e89d2e00a78fd37a0e08c48ef91740aec0175" not found
E1214 22:48:55.543248       1 csi_attacher.go:202] kubernetes.io/csi: attacher.VolumesAreAttached failed for attach.ID=csi-27bdc8eefc86cffa9c3b14fabe1e89d2e00a78fd37a0e08c48ef91740aec0175: volumeattachments.storage.k8s.io "csi-27bdc8eefc86cffa9c3b14fabe1e89d2e00a78fd37a0e08c48ef91740aec0175" not found
E1214 22:48:55.892983       1 csi_attacher.go:202] kubernetes.io/csi: attacher.VolumesAreAttached failed for attach.ID=csi-27bdc8eefc86cffa9c3b14fabe1e89d2e00a78fd37a0e08c48ef91740aec0175: volumeattachments.storage.k8s.io "csi-27bdc8eefc86cffa9c3b14fabe1e89d2e00a78fd37a0e08c48ef91740aec0175" not found
I1214 22:49:26.043773       1 operation_generator.go:359] AttachVolume.Attach succeeded for volume "hosting-vol-data-mariadb-hc1-md42-1" (UniqueName: "kubernetes.io/csi/linstor.csi.linbit.com^hosting-vol-data-mariadb-hc1-md42-1") from node "m11c15" 
I1214 22:49:26.043872       1 event.go:281] Event(v1.ObjectReference{Kind:"Pod", Namespace:"hosting", Name:"hc1-md42-s1-0", UID:"4f865b89-49ba-4565-9285-7e6ed1e0ed14", APIVersion:"v1", ResourceVersion:"2114111218", FieldPath:""}): type: 'Normal' reason: 'SuccessfulAttachVolume' AttachVolume.Attach succeeded for volume "hosting-vol-data-mariadb-hc1-md42-1" 
```
Due backoff loop the container is starting only after 10 minutes.

**What you expected to happen**:

* controller-manager detaches volume from the old node immideately after the pod is removed and attaches it to new one immediately after detaching.
* have no these messages:
```
csi_attacher.go:202] kubernetes.io/csi: attacher.VolumesAreAttached failed for attach.ID=csi-27bdc8eefc86cffa9c3b14fabe1e89d2e00a78fd37a0e08c48ef91740aec0175: volumeattachments.storage.k8s.io "csi-27bdc8eefc86cffa9c3b14fabe1e89d2e00a78fd37a0e08c48ef91740aec0175" not 
```

**How to reproduce it (as minimally and precisely as possible)**:

* Create StatefulSet
* Run `kubectl get pod -o wide | grep <statefulsetname>` for find the node where is it running 
* `kubectl cordon <node>`
* `kubectl delete <podname>`
* Run `kubectl get pod -o wide | grep <statefulsetname>` for find the new node is it running 

Container will stuck for 10 minutes in **ContainerCreating** state

You can also check `kubectl get volumeattachments.storage.k8s.io -w` for track detaching/attaching process.

**Anything else we need to know?**:

Yes, we have two clusters, staging cluster have no this problem unlike production one.

**Environment**:
- Kubernetes version:
```
Client Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.0", GitCommit:"70132b0f130acc0bed193d9ba59dd186f0e634cf", GitTreeState:"clean", BuildDate:"2019-12-07T21:20:10Z", GoVersion:"go1.13.4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.0", GitCommit:"70132b0f130acc0bed193d9ba59dd186f0e634cf", GitTreeState:"clean", BuildDate:"2019-12-07T21:12:17Z", GoVersion:"go1.13.4", Compiler:"gc", Platform:"linux/amd64"}
```
- CSI version:
```
quay.io/k8scsi/csi-provisioner:v1.4.0
quay.io/k8scsi/csi-attacher:v2.0.0
quay.io/k8scsi/csi-snapshotter:v1.2.2
quay.io/k8scsi/csi-node-driver-registrar:v1.2.0
```
- Cloud provider or hardware configuration: Bare Metal
- OS: Ubuntu 18.04.2 LTS
- Kernel `4.15.18-18-pve`
- Install tools: kubeadm
- CSI plugin and version: [`linstor-csi v0.7.2`](https://github.com/LINBIT/linstor-csi)

- Others:

**Comment by msau42** [comment]
@kubernetes/sig-storage-bugs

**Comment by dimm0** [comment]
CSI volumes are taking from 10 minutes to never to attach after upgrading to 1.17.1

**Comment by dimm0** [comment]
rook cephFS driver

**Comment by dimm0** [comment]
workqueue_adds_total volumes metric for controller in prometheus shows 85 ops for "volumes", highest of all. What does it mean?

**Comment by dimm0** [comment]
Mine suddenly started working

**Comment by gnufied** [comment]
What does the metric for attach and detach look like?

**Comment by dimm0** [comment]
Controller manager "volumes" and "claims" add rate is still at 86 ops
work queue latency shows 9s for "disruption"

Scheduler latency still shows 17 minutes for volume, I'm surprised! Maybe haven't reset yet

**Comment by gnufied** [comment]
ack. I am interested in attach_detach operation metrics from controller-manager. They should tell us more accurately if it is indeed attach and detach operation that is taking long time.

**Comment by dimm0** [comment]
https://grafana.nautilus.optiputer.net/d/72e0e05bef5099e5f049b05fdc429ed4/kubernetes-controller-manager?orgId=1
See it here?

**Comment by dimm0** [comment]
Seems like it's not scaling with the number of volumes...
With 300 volumes back, it's slow again

**Comment by dimm0** [comment]
@gnufied we're stuck with jobs in the cluster.. anything I can do to help debugging this?

**Comment by jsafrane** [comment]
From the attacher logs it seems that it took the CSI driver itself ~2 minutes to detach the volume:

```
I1214 22:45:19.493975       1 operation_generator.go:1297] Verified volume is safe to detach for volume "hosting-vol-data-mariadb-hc1-md42-1" (UniqueName: "kubernetes.io/csi/linstor.csi.linbit.com^hosting-vol-data-mariadb-hc1-md42-1") on node "m8c29" 
I1214 22:47:09.893665       1 operation_generator.go:470] DetachVolume.Detach succeeded for volume "hosting-vol-data-mariadb-hc1-md42-1" (UniqueName: "kubernetes.io/csi/linstor.csi.linbit.com^hosting-vol-data-mariadb-hc1-md42-1") on node "m8c29" 
```
The same for volume attachment:
```
I1214 22:47:09.924794       1 reconciler.go:288] attacherDetacher.AttachVolume started for volume "hosting-vol-data-mariadb-hc1-md42-1" (UniqueName: "kubernetes.io/csi/linstor.csi.linbit.com^hosting-vol-data-mariadb-hc1-md42-1") from node "m11c15" 
I1214 22:49:26.043773       1 operation_generator.go:359] AttachVolume.Attach succeeded for volume "hosting-vol-data-mariadb-hc1-md42-1" (UniqueName: "kubernetes.io/csi/linstor.csi.linbit.com^hosting-vol-data-mariadb-hc1-md42-1") from node "m11c15" 
```
There is very little that Kubernetes can do. Check with rook.

**Comment by kvaps** [comment]
@jsafrane volume attachment is working fine with **controller-manager v1.16.4** and very slow **controller-manager v1.17.0**, what the difference?

> it took the CSI driver itself ~2 minutes

No, CSI driver is detaching volume almost immediately if do `kubectl delete volumeattachment` manually.
The same about the attaching, driver is handling new volumeattachment requests quite fast. 

The problem is that controller-manager does not creating these requests in a time.

**Comment by dimm0** [comment]
@jsafrane Same here as what @kvaps wrote.

Also I found that it's not only for rook volumes, we also use CVMFS CSI driver from CERN, same thing. And in our cluster the problem appears when the number of volumes attached (or requested to attach) is in hundreds (CVMFS needs 5-6 volumes per pod attached, but it worked fine in 16.3)

**Comment by kvaps** [comment]
I'm using Linstor not Rook

**Comment by humblec** [comment]
@kvaps ooc, do you observe the problem for other "deployment" kinds too ? or specific to "sts" ?

**Comment by kvaps** [comment]
> @kvaps ooc, do you observe the problem for other "deployment" kinds too ? or is it specific to "sts" ?

Yes I have two clusters, one is small and it does not having this problem, it is containing: 12 nodes and 145 PV

And I have larger one, where this problem persist, it is containing: 543 node and 673 PV

**Comment by dimm0** [comment]
Anybody tried to confirm the issue at least?

**Comment by raunakkumar** [comment]
Seeing the issue on our setup too with k8s 1.17 and CSI volumes

Following logs are seen in the controller manager
```
E0221 01:42:05.304849       1 csi_attacher.go:202] kubernetes.io/csi: attacher.VolumesAreAttached failed for attach.ID=csi-7396ecd93940b6f44840b6376ae7fbe2567030097301971fae1f51a3830050eb: volumeattachments.storage.k8s.io "csi-7396ecd93940b6f44840b6376ae7fbe2567030097301971fae1f51a3830050eb" not found
E0221 01:42:05.554025       1 csi_attacher.go:202] kubernetes.io/csi: attacher.VolumesAreAttached failed for attach.ID=csi-7396ecd93940b6f44840b6376ae7fbe2567030097301971fae1f51a3830050eb: volumeattachments.storage.k8s.io "csi-7396ecd93940b6f44840b6376ae7fbe2567030097301971fae1f51a3830050eb" not found
E0221 01:42:05.604643       1 csi_attacher.go:202] kubernetes.io/csi: attacher.VolumesAreAttached failed for attach.ID=csi-7396ecd93940b6f44840b6376ae7fbe2567030097301971fae1f51a3830050eb: volumeattachments.storage.k8s.io "csi-7396ecd93940b6f44840b6376ae7fbe2567030097301971fae1f51a3830050eb" not found
```

**Comment by sfiligoi** [comment]
This is basically making our Kubernetes cluster unusable.
Any ETA on when it will be fixed?

**Comment by msau42** [comment]
cc @davidz627 
I believe there were some changes around the verified volumes attached area recently.

**Comment by tedyu** [comment]
@kvaps @raunakkumar 
Do you see any of the logs added by:
https://github.com/kubernetes/kubernetes/pull/83593/files

I assume the above was what @msau42 referred to.

**Comment by raunakkumar** [comment]
Hi @tedyu Ours is not an in-tree plugin or the cloud provider one. 
Upgrading to csi-attacher v2.1.1 and setting the `-reconcile-sync`  to a larger value definitely improved our situation as indicated by @saad-ali at https://github.com/kubernetes/kubernetes/issues/84169#issuecomment-545692734.

**Comment by msau42** [comment]
https://github.com/kubernetes/kubernetes/pull/84321 was made in 1.17 and increased the detach timeout from 15 seconds to 2 mins. That could potentially be affecting the issue seen here

**Comment by tedyu** [comment]
Looking at the code where the attachID surfaced:
```
		attach, err := c.k8s.StorageV1().VolumeAttachments().Get(context.TODO(), attachID, meta.GetOptions{})
		if err != nil {
			attached[spec] = false
			klog.Error(log("attacher.VolumesAreAttached failed for attach.ID=%v: %v", attachID, err))
```
It seems csiAttacher should have some mechanism of not querying for attachID's which are not found - some type of throttling.
I doubt that the attachID's mentioned in @raunakkumar 's comment above would ever be found.

@saad-ali @msau42 @davidz627 
What do you think ?

**Comment by kvaps** [comment]
> #84321 was made in 1.17 and increased the detach timeout from 15 seconds to 2 mins. That could potentially be affecting the issue seen here

Just checked that reverting b4e0f1b8f8b4de2fadf7a875742a2da913cd0a27 from #84321 does not solves the original problem.

Problem still persist on v1.17 and v1.18 versions

**Comment by kvaps** [comment]
But after upgrade to v1.18 I also see many `Throttling request took` messages:

```
I0429 16:32:19.094692       1 reconciler.go:203] attacherDetacher.DetachVolume started for volume "pvc-e45016a9-03e0-4fa1-a17e-18e289f34080" (UniqueName: "kubernetes.io/csi/linstor.csi.linbit.com^pvc-e45016a9-03e0-4fa1-a17e-18e289f34080") on node "m11c2" 
I0429 16:32:25.877871       1 request.go:621] Throttling request took 47.696514122s, request: GET:https://10.36.20.0:6443/apis/storage.k8s.io/v1/volumeattachments/csi-28fc1e2725b77d70ec51d15560021a1b964566ae424c5bd7647b32b753d4298b
I0429 16:32:35.880547       1 request.go:621] Throttling request took 47.49928719s, request: GET:https://10.36.20.0:6443/apis/storage.k8s.io/v1/volumeattachments/csi-1ddae3d436302c1089fc5e6e9f9ca0af09c8932f58bc6e5cf96a88e39f917af1
I0429 16:32:45.928172       1 request.go:621] Throttling request took 44.187281711s, request: GET:https://10.36.20.0:6443/apis/storage.k8s.io/v1/volumeattachments/csi-adaf0e809b7cc0586bc224b7e344b4d845990a3d7f7151abda66069a5b79c71f
I0429 16:32:54.933247       1 operation_generator.go:1384] Verified volume is safe to detach for volume "pvc-e45016a9-03e0-4fa1-a17e-18e289f34080" (UniqueName: "kubernetes.io/csi/linstor.csi.linbit.com^pvc-e45016a9-03e0-4fa1-a17e-18e289f34080") on node "m11c2" 
I0429 16:32:55.977920       1 request.go:621] Throttling request took 36.877943129s, request: GET:https://10.36.20.0:6443/apis/storage.k8s.io/v1/volumeattachments/csi-a0a8584493e27052fb5e72bbd9d66621d0d8152bd8866b97882f6e66d5dff172
I0429 16:33:05.981974       1 request.go:621] Throttling request took 46.867869348s, request: GET:https://10.36.20.0:6443/apis/storage.k8s.io/v1/volumeattachments/csi-adaf0e809b7cc0586bc224b7e344b4d845990a3d7f7151abda66069a5b79c71f
E0429 16:33:11.586607       1 csi_attacher.go:204] kubernetes.io/csi: attacher.VolumesAreAttached failed for attach.ID=csi-f668145654a032f85932a6f09b1af40467bf89771d8420a3315058695cd76daf: volumeattachments.storage.k8s.io "csi-f668145654a032f85932a6f09b1af40467bf89771d8420a3315058695cd76daf" not found
E0429 16:33:11.838629       1 csi_attacher.go:204] kubernetes.io/csi: attacher.VolumesAreAttached failed for attach.ID=csi-f668145654a032f85932a6f09b1af40467bf89771d8420a3315058695cd76daf: volumeattachments.storage.k8s.io "csi-f668145654a032f85932a6f09b1af40467bf89771d8420a3315058695cd76daf" not found
I0429 16:33:16.028513       1 request.go:621] Throttling request took 56.902526422s, request: GET:https://10.36.20.0:6443/apis/storage.k8s.io/v1/volumeattachments/csi-afe39e8a371598ea617aa9066b555079e0801aeef5ca9b34d25fc82dc1dc9874
I0429 16:33:26.078118       1 request.go:621] Throttling request took 1m3.4408186s, request: GET:https://10.36.20.0:6443/apis/storage.k8s.io/v1/volumeattachments/csi-330286d1baf8860491c8ab44a38bfb62f591e45a73ecf142f5d0270c15e59be9
I0429 16:33:36.127925       1 request.go:621] Throttling request took 1m2.287143818s, request: GET:https://10.36.20.0:6443/apis/storage.k8s.io/v1/volumeattachments/csi-7990c6741426fb01ec7f65ad2709cc69bcdd764357ec816370cf475e03f801bf
I0429 16:33:46.180751       1 request.go:621] Throttling request took 1m1.291943967s, request: GET:https://10.36.20.0:6443/apis/storage.k8s.io/v1/volumeattachments/csi-b24c87b455141429422c21eb72a7f8fb7388067f7c04ab76f6965ffac76ad02f
I0429 16:33:53.149316       1 operation_generator.go:472] DetachVolume.Detach succeeded for volume "pvc-e45016a9-03e0-4fa1-a17e-18e289f34080" (UniqueName: "kubernetes.io/csi/linstor.csi.linbit.com^pvc-e45016a9-03e0-4fa1-a17e-18e289f34080") on node "m11c2" 
I0429 16:33:53.186509       1 reconciler.go:275] attacherDetacher.AttachVolume started for volume "pvc-e45016a9-03e0-4fa1-a17e-18e289f34080" (UniqueName: "kubernetes.io/csi/linstor.csi.linbit.com^pvc-e45016a9-03e0-4fa1-a17e-18e289f34080") from node "m11c1" 
I0429 16:33:56.180766       1 request.go:621] Throttling request took 58.135944699s, request: GET:https://10.36.20.0:6443/apis/storage.k8s.io/v1/volumeattachments/csi-483202fcefd5aac50cb4b6c6ccea397a053fa45d64afc1d157584d14d0613da7
I0429 16:34:06.184419       1 request.go:621] Throttling request took 57.848165904s, request: GET:https://10.36.20.0:6443/apis/storage.k8s.io/v1/volumeattachments/csi-d43ac28efa67a016d4d0bcbbb08537548d1fae906e8b7e0db4939d9a138cabcc
I0429 16:34:16.234609       1 request.go:621] Throttling request took 57.601839534s, request: GET:https://10.36.20.0:6443/apis/storage.k8s.io/v1/volumeattachments/csi-1d01535127f6c1b4b86738af2fb9f13d146716be959167df92b2b2ae69c6b04b
I0429 16:34:26.282358       1 request.go:621] Throttling request took 1m7.112379532s, request: GET:https://10.36.20.0:6443/apis/storage.k8s.io/v1/volumeattachments/csi-038291e857c76f0b849f9fb19daafe08a65db32b7a21b0fd3107f4f9993cc99c
I0429 16:34:36.328420       1 request.go:621] Throttling request took 1m17.146054029s, request: GET:https://10.36.20.0:6443/apis/storage.k8s.io/v1/volumeattachments/csi-e85f15e04fd1c4368d4f80539015fc3f54ef326b263493d9f633989782a846e3
I0429 16:34:46.377909       1 request.go:621] Throttling request took 1m25.496478794s, request: GET:https://10.36.20.0:6443/apis/storage.k8s.io/v1/volumeattachments/csi-be0e21dee5bd9fd76915bc67840935c6ca51d5db5f6d1196b9e2b5146b513390
I0429 16:34:56.427903       1 request.go:621] Throttling request took 1m25.094061943s, request: GET:https://10.36.20.0:6443/apis/storage.k8s.io/v1/volumeattachments/csi-a8e6348e2b471271b4529caeabd8fde627f691155982cf1f6ffe235593ffa0d2
I0429 16:35:06.432514       1 request.go:621] Throttling request took 1m24.620478627s, request: GET:https://10.36.20.0:6443/apis/storage.k8s.io/v1/volumeattachments/csi-b3d73261eeb3420c540eb2f9d6d1f743f33022898f0fd733ea96ccbe82bb9699
I0429 16:35:16.477894       1 request.go:621] Throttling request took 1m19.04497522s, request: GET:https://10.36.20.0:6443/apis/storage.k8s.io/v1/volumeattachments/csi-939e15603bf43c945b2e5918a98c8989a924bfe59329850257b25ddaecf9533b
I0429 16:35:26.528480       1 request.go:621] Throttling request took 1m17.979692009s, request: GET:https://10.36.20.0:6443/apis/storage.k8s.io/v1/volumeattachments/csi-110ed8a76d8caa7e581041fb443f05c2ac69de6c704efbbf5bd41b5ed0dd02fb
I0429 16:35:36.577882       1 request.go:621] Throttling request took 1m17.365090427s, request: GET:https://10.36.20.0:6443/apis/storage.k8s.io/v1/volumeattachments/csi-1791c0a48fcb783ce1bcef3211b9cbdd798d92e718d3182fda2dfd76c73214e3
I0429 16:35:46.577990       1 request.go:621] Throttling request took 1m27.346257533s, request: GET:https://10.36.20.0:6443/apis/storage.k8s.io/v1/volumeattachments/csi-a0a8584493e27052fb5e72bbd9d66621d0d8152bd8866b97882f6e66d5dff172
I0429 16:35:56.581232       1 request.go:621] Throttling request took 1m37.33178429s, request: GET:https://10.36.20.0:6443/apis/storage.k8s.io/v1/volumeattachments/csi-c176a73a13a7b4836f363018c5a0834c457d842686e15128cd99a1db34a509b8
I0429 16:36:06.627938       1 request.go:621] Throttling request took 1m44.748045488s, request: GET:https://10.36.20.0:6443/apis/storage.k8s.io/v1/volumeattachments/csi-2002f0bc0bbd0dbeeef422cdece720ffb2953a1e98e1a5f47a254372154e3446
E0429 16:36:13.783387       1 csi_attacher.go:204] kubernetes.io/csi: attacher.VolumesAreAttached failed for attach.ID=csi-f668145654a032f85932a6f09b1af40467bf89771d8420a3315058695cd76daf: volumeattachments.storage.k8s.io "csi-f668145654a032f85932a6f09b1af40467bf89771d8420a3315058695cd76daf" not found
I0429 16:36:16.677903       1 request.go:621] Throttling request took 1m44.796856167s, request: GET:https://10.36.20.0:6443/apis/storage.k8s.io/v1/volumeattachments/csi-8becff6d271e97e90f21c00af0b54062c64a3f50c9fd4bf6384d34b6d31bb9ee
E0429 16:36:17.189106       1 csi_attacher.go:204] kubernetes.io/csi: attacher.VolumesAreAttached failed for attach.ID=csi-f668145654a032f85932a6f09b1af40467bf89771d8420a3315058695cd76daf: volumeattachments.storage.k8s.io "csi-f668145654a032f85932a6f09b1af40467bf89771d8420a3315058695cd76daf" not found
I0429 16:36:26.727914       1 request.go:621] Throttling request took 1m44.346001622s, request: GET:https://10.36.20.0:6443/apis/storage.k8s.io/v1/volumeattachments/csi-e6d3d69e9b80ae2aae36f69cf000afefbed961fe8a7d2c110bae9522d65d59c9
I0429 16:36:36.777924       1 request.go:621] Throttling request took 1m37.087458713s, request: GET:https://10.36.20.0:6443/apis/storage.k8s.io/v1/volumeattachments/csi-7adcdd16bf77644bd6ec3574bea5a9efe0735373572bf3659154d118a8063b32
I0429 16:36:42.638519       1 operation_generator.go:361] AttachVolume.Attach succeeded for volume "pvc-e45016a9-03e0-4fa1-a17e-18e289f34080" (UniqueName: "kubernetes.io/csi/linstor.csi.linbit.com^pvc-e45016a9-03e0-4fa1-a17e-18e289f34080") from node "m11c1" 
I0429 16:36:42.638768       1 event.go:278] Event(v1.ObjectReference{Kind:"Pod", Namespace:"default", Name:"mypod", UID:"2bdd00e4-675e-4331-836d-9c4b8731c17b", APIVersion:"v1", ResourceVersion:"3352539840", FieldPath:""}): type: 'Normal' reason: 'SuccessfulAttachVolume' AttachVolume.Attach succeeded for volume "pvc-e45016a9-03e0-4fa1-a17e-18e289f34080" 
I0429 16:36:46.778407       1 request.go:621] Throttling request took 1m29.997244022s, request: GET:https://10.36.20.0:6443/apis/storage.k8s.io/v1/volumeattachments/csi-77864ad9e8ec8b356b47b489c05a4ac9c00d7642f7149c5fbb5ca5b8cde03182
I0429 16:36:56.827864       1 request.go:621] Throttling request took 1m37.540315237s, request: GET:https://10.36.20.0:6443/apis/storage.k8s.io/v1/volumeattachments/csi-8b9c8ca320838154169daa41e1782fc816668123a22565d6afb282fc282f50e2
I0429 16:37:06.827900       1 request.go:621] Throttling request took 1m47.532758877s, request: GET:https://10.36.20.0:6443/apis/storage.k8s.io/v1/volumeattachments/csi-0bfbfa96086497f64a9de8d8bfc4afd7d3b79349246a223b85fc3825e2d48f27
I0429 16:37:16.828522       1 request.go:621] Throttling request took 1m57.521223191s, request: GET:https://10.36.20.0:6443/apis/storage.k8s.io/v1/volumeattachments/csi-67f84beaf556dcb8089b760e3ced2da94313877b9d201c33fe1ba22d0e93ae72
I0429 16:37:26.881003       1 request.go:621] Throttling request took 1m57.249509892s, request: GET:https://10.36.20.0:6443/apis/storage.k8s.io/v1/volumeattachments/csi-ea06098d8a624fcccb072039c424316ccbfda35303087c858120b0b385c9767f
I0429 16:37:36.927883       1 request.go:621] Throttling request took 1m56.64731948s, request: GET:https://10.36.20.0:6443/apis/storage.k8s.io/v1/volumeattachments/csi-5791cd76e78a65ef5970d96eaa80ed02f39fc1481816b6c811e67307e208a8ec
I0429 16:37:46.927951       1 request.go:621] Throttling request took 1m56.34687169s, request: GET:https://10.36.20.0:6443/apis/storage.k8s.io/v1/volumeattachments/csi-0ae6cd8ef2b4310bb1eabd12685abad0a6ce2677e6815d44f2c4767fbc00106a
I0429 16:37:56.977972       1 request.go:621] Throttling request took 1m56.093840598s, request: GET:https://10.36.20.0:6443/apis/storage.k8s.io/v1/volumeattachments/csi-3d9bfa0c16dd5abde71fdfea604c5432e88affb1f3b3a967c095dc8e1e1cd905
```

**Comment by msau42** [comment]
Do you have a lot of volumes in your cluster? You may be hitting this scaling issue: https://github.com/kubernetes/kubernetes/issues/84169

**Comment by kvaps** [comment]
@msau42, thanks for pointing
I have 803 PVs, 588 of them are CSI

If I downgrade controller-manager to v1.16.9, reattachment is working fine

**Comment by cpointner** [comment]
We experience the exact same problem using Netapp/trident. Ever since we upgraded to Kubernetes 1.17.5 (from 1.16.8) attaching volumes to nodes takes up to 30 Minutes. The cluster is not very big - all in all we have less then 100 PVs and only a few PVCs get created/deleted every day. Also the problem can be reproduced by adding a single PVC and starting one pod that only uses this one new PVC.

As others already mentioned the step that takes the most time is the creation of the VolumeAttachment object. In our case once the pod gets created the controler manager more or less immediately logs this:
```
I0430 14:37:41.231375       1 operation_generator.go:359] AttachVolume.Attach succeeded for volume "pvc-02f957fd-626e-4be7-9284-ab944c54f922" (UniqueName: "kubernetes.io/csi/csi.trident.netapp.io^pvc-02f957fd-626e-4be7-9284-ab944c54f922") from node "worker2"
```
which tells me that it correctly detected that there is a volume that needs to be attached using trident. Alas the corresponding VolumeAttachment object does not get created by the controller-manager. After about 30 minutes this object is finally created and trident immediately attaches the volume. After this it again takes minutes until the kubelet starts the pod which is most likely caused by exponential backoff. When the pod is stopped detaching the volume also takes very long. Again this is because the controller-manager does not try to delete the VolumeAttachment for quite while. Once the Object gets deleted trident immediately detaches the volume which finally allows the PV to be deleted.

What works for us is to just kill the leading controller-manager. Once one of the other controller managers takes over attaching and detaching the volume is more ore less instant. But this fix is not permanent because after some time the other controller-manager behaves the same until it again gets restarted.

As already said this problem appeared after an upgrade from Kubernetes 1.16.8 to 1.17.5. This upgrade was accompanied by an upgrade of trident from 19.10.0 to 20.01.1 (which is needed to work with Kubernetes 1.17).  Interestingly we run other Kubernetes Clusters that received exactly the same upgrade which so far don't show any problems.

**Comment by cpointner** [comment]
I'm now more sure then ever that this is in fact a but in controller-manager. See the attached image which shows the number of gorotines as reported by the controller-manager via prometheus metrics:

![goroutines](https://user-images.githubusercontent.com/29884944/81068235-50f0d380-8ee0-11ea-9972-be1864890342.png)

As you can see here the number of go-routines is constantly growing over time. At the end of the graph you can see what happens if i restart the leading controller manager. One of the remaining controllers takes over and immediately starts to pile up go-routines. I also enable profiling and pulled the following stack dump:

[pprof-goroutines.dump.gz](https://github.com/kubernetes/kubernetes/files/4580863/pprof-goroutines.dump.gz)

**Comment by cpointner** [comment]
In comparison this is how the same graph looks like at a different cluster where we don't experience the same problem. This second cluster is using the same version of Kubernetes and Trident:

![goroutines-healthy](https://user-images.githubusercontent.com/29884944/81069521-48999800-8ee2-11ea-9a97-792272f87de2.png)

**Comment by kvaps** [comment]
But in this way it should handle attach/detach operations  faster in the first time rather then later.
Regarding to my case it is exact the same time. It's not getting worse during the time spent while controller-manager is running.

I found out that problem persists on controller-manager `v1.17.0` and higher versions but not persists on `v1.16.9`

**Comment by cpointner** [comment]
Yeah i think you are right. I will open up a new issue.

**Comment by msau42** [comment]
This change to client-go that enabled throttling for erroneously unthrottled clients probably explains why we see increased throttling of controller-manager in 1.17: https://github.com/kubernetes/kubernetes/pull/86168

I think the fix is the same. We should change the VerifyVolumesAttached() method to use informers instead of doing a GET on every volumeattachment object every minute. @yuga711 is working on a fix.

**Comment by kvaps** [comment]
Just short note that only v1.18 have throttling messages, v1.17 have no them.

**Comment by cpointner** [comment]
Thanks @msau42 and @yuga711 this sounds good. I think that this will help with our problem. Anyway i still think that there is a problem with the controller manager piling up so many go-routines. Based on the information i gathered so far it has maybe something to do with the horizontal-pod-autoscaler. Of course this is outside of the scope of this issue. I will open up a new one for that as soon as i have time to finish my analysis.

**Comment by zhucan** [comment]
> This change to client-go that enabled throttling for erroneously unthrottled clients probably explains why we see increased throttling of controller-manager in 1.17: #86168
> 
> I think the fix is the same. We should change the VerifyVolumesAttached() method to use informers instead of doing a GET on every volumeattachment object every minute. @yuga711 is working on a fix.

When to solve this problem?

**Comment by yuga711** [comment]
In progress. Planning to submit to 1.19 and then backport to earlier releases.

**Comment by manojvs157** [comment]
Will this be patched to 1.17 as well ?

**Comment by msau42** [comment]
Duplicating this issue to #84169
/close

**Comment by yeganeahmadnejad** [comment]
hi 
because of https://github.com/kubernetes/kubernetes/issues/90589 issue I had to upgrade my kube cluster form 16.4 to 18.2
and have same problem
I0622 05:36:20.336626       1 request.go:621] Throttling request took 3h32m8.394756742s, request: GET:https://192.168.100.68:6443/apis/storage.k8s.io/v1/volumeattachments/csi-3dee2b16f91302d22fe3b48c80316885378019fa46ddb38eaede2a8454f8a211

**Comment by yuga711** [comment]
> because of #90589 issue I had to upgrade my kube cluster form 16.4 to 18.2
> and have same problem

I think [the PR](https://github.com/kubernetes/kubernetes/pull/92015) has not been merged into 1.18 yet?

**Comment by kvaps** [comment]
This one was probably solved, I have no problems running it on `v1.18.2`
